### PR TITLE
Add support for field aliases to 6.x.

### PR DESCRIPTION
**Pull request by jtibshirani**
This backport is composed of the commits from the original PR (#32172), three backported bugfixes, plus three new commits that address issues specific to 6.x. The first batch of commits are direct cherry-picks from master, with minor conflict resolution. The additional commits come last, and don't have links to previous PRs in their title.

The backport went smoothly overall, but there were some conflicts relating to the fact that in 6.x., multiple mapping types can be present.

### CODE REVIEW

**Comment by jtibshirani** [comment]
@elasticmachine test this please

**Comment by jtibshirani** [comment]
This backport is composed of all the commits from the master PR (#32172), plus three additional commits that address issues specific to 6.x. These additional commits come last, and don't include links to a previous PR in their title.

The backport went smoothly overall, but there were some conflicts I needed to resolve relating to the fact that in 6.x, multiple mapping types can be present.

**Comment by jtibshirani** [comment]
@colings86 @jpountz this is ready for review.

**Comment by jtibshirani** [comment]
@elasticmachine test this please

**Review by jpountz — COMMENTED** [review]
I'm not a fan of the fact that we need more complex logic to handle the fact that there may be more types. Should we consider only allowing aliases on 6.x indices so that we don't have to deal with this case and have exactly the same logic as in master?

**Comment by colings86** [comment]
> I'm not a fan of the fact that we need more complex logic to handle the fact that there may be more types. Should we consider only allowing aliases on 6.x indices so that we don't have to deal with this case and have exactly the same logic as in master?

Personally I would be fine with that trade-off

File: docs/reference/mapping/types/alias.asciidoc [hunk 1]
```diff
@@ -0,0 +1,105 @@
+[[alias]]
+=== Alias datatype
+
+NOTE: Field aliases can only be specified on indexes with a single mapping type. To add a field
+alias, the index must therefore have been created in 6.0.0 or later, or be an older index with
+the setting `index.mapping.single_type: true`. Please see <<removal-of-types>> for more information.
```
Comment (jtibshirani) [inline docs/reference/mapping/types/alias.asciidoc]: I wasn't sure whether we should accept this setting, or if we should strictly enforce that the index be created in 6.x or later. I'm concerned that there will be confusing behavior in the following case: this setting is enabled, a field alias is added, then the setting is disabled.

**Comment by jtibshirani** [comment]
Great, thank you! I've removed the extra logic and added validation to ensure that the index has a single type. I left a question above about the approach, and would appreciate if someone could take a look.

File: docs/reference/mapping/types/alias.asciidoc [hunk 2]
```diff
@@ -0,0 +1,105 @@
+[[alias]]
+=== Alias datatype
+
+NOTE: Field aliases can only be specified on indexes with a single mapping type. To add a field
+alias, the index must therefore have been created in 6.0 or later, or be an older index with
+the setting `index.mapping.single_type: true`. Please see <<removal-of-types>> for more information.
```
Comment (jtibshirani) [inline docs/reference/mapping/types/alias.asciidoc]: I wasn't sure whether we should accept this setting, or if we should strictly enforce that the index be created in 6.0 or later. I'm concerned that there will be confusing behavior in the following case: this setting is enabled, a field alias is added, then the setting is disabled.

**Review by jpountz — APPROVED** [review]
Thanks, it looks good to me. I'm wondering whether only allowing on 6.x rather than 6.x plus single-type 5.x indices would be less confusing. For instance right now error messages mention that aliases can't be added because there are multiple types, which is not exact. The `single_type` setting is something that users opt in at index creation time that tells Elasticsearch to reject mapping updates that would create more than one type. But if a 5.x index was created without this setting, then `IndexSettings#isSingleType` will return false even if there is only one type in the index.

**Comment by jtibshirani** [comment]
@elasticmachine test this please

**Comment by jtibshirani** [comment]
@jpountz I'm going to merge this, but will give it more thought and follow-up with a change that either checks only for 6.x indexes, or at least tightens the wording of error messages.

**Comment by jpountz** [comment]
Sounds good to me. Thanks @jtibshirani !

**Comment by jtibshirani** [comment]
An update: I opened #32482 to fix the error-message wording. I decided to stick with the current check for a couple reasons:
- For simplicity, it’s nice to keep parity with the `join` field in terms of when the datatypes can be used.
- It provides some flexibility in terms of being able to try out the feature, at very little cost in terms of code complexity.

**Pull request merged**